### ISSUE TITLE
[Bug]: connected_components(): float16 loses integer precision above 2048, causing label ID collisions

### ISSUE THREAD

**Issue by DPug888** [kornia/kornia#3864]
### 🐛 Describe the bug

`kornia.contrib.connected_components` generates internal per-pixel ID numbers 
using the *input image's own dtype*. When the input is `float16` and the image 
has more than 2,048 pixels (true for almost any real image), two genuinely 
disconnected components can be silently assigned the **same label** — with no 
error, warning, or crash.

Labels are initialized here:
```python
out = torch.arange(1, B * H * W + 1, device=image.device, dtype=image.dtype).view((-1, 1, H, W))
```
`float16` can represent whole numbers exactly only up to 2¹¹ = 2048 (10 mantissa 
bits + 1 implied bit). Beyond that, it must round to the nearest representable 
value, so distinct integer IDs collide onto the same float16 value. If two 
disconnected components receive colliding starting IDs, they end up reporting 
the same final label after the max-pooling propagation.

### 🔄 Steps to Reproduce

```bash
1. Create a binary image (H=3000, W=8) with two isolated single-pixel foreground regions, 2 rows apart (spatially disconnected).
2. Run connected_components() on the image cast to torch.float32.
3. Run connected_components() on the same image cast to torch.float16.
4. Compare the labels assigned to the two pixels in each case.
```

### 💻 Minimal Code Example

```python
import torch
from kornia.contrib import connected_components

W, H = 8, 3000
img = torch.zeros(1, 1, H, W)
img[0, 0, 2051, 0] = 1.0   # component A
img[0, 0, 2053, 0] = 1.0   # component B, 2 rows away -> NOT connected

labels_f32 = connected_components(img.to(torch.float32), num_iterations=150)
labels_f16 = connected_components(img.to(torch.float16), num_iterations=150)

print(labels_f32[0,0,2051,0].item(), labels_f32[0,0,2053,0].item())
# -> 16409.0 16425.0  (correct: different labels)

print(labels_f16[0,0,2051,0].item(), labels_f16[0,0,2053,0].item())
# -> 16416.0 16416.0  (bug: identical label despite being disconnected)
```

### ✅ Expected behavior

Two spatially disconnected foreground pixels/regions should always receive different labels, regardless of the input tensor's dtype.

### ❌ Actual behavior

When the input dtype is float16, connected_components() assigns the same label to two genuinely disconnected pixels. No error or warning is raised. This happens because internal label IDs are generated with torch.arange(..., dtype=image.dtype), and float16 can only represent integers exactly up to 2048 — beyond that, distinct IDs round to the same float16 value and collide.

### 🔧 Environment

```shell
- Kornia version: 0.8.3
- PyTorch version: 2.9.1+cpu
- Python version: 3.13.9
- OS (e.g., Linux, macOS, Windows): Windows
- Installation method (pip, conda, from source): from source
- CUDA/cuDNN version (if applicable):NA
- GPU model (if applicable):NA

You can also run:

wget https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

### 🤝 Contribution Intent

- [x] I plan to submit a PR to fix this bug
- [ ] I'm reporting this bug but not planning to fix it

**Comment by ahmadmasood43** [comment]
Hi! I'd like to work on this. could you assign it to me?

I've root-caused and fixed it: the flood-fill algorithm labels each pixel with its linear index, but the whole iterative max-pooling loop ran in image.dtype. Since float16 only represents integers exactly up to 2¹¹ = 2048, this wasn't just a cosmetic labeling issue. the flood fill propagation itself corrupted for images with more than ~2048 pixels, causing genuinely disconnected components to merge.


Fix (in kornia/contrib/connected_components.py):

Run the iterative computation internally in float32 (or float64 if the input already is that), regardless of image.dtype, so the flood-fill never loses precision during propagation.
After convergence, remap the pixel-index-based labels to compact consecutive integers (0 for background, 1..N for the N components) via torch.unique(..., return_inverse=True), then cast back to image.dtype. This matches how OpenCV/scikit-image label components, and makes the final cast to a low-precision dtype safe regardless of image size, as long as there are under 2048 components (essentially always true in practice).
Also updated the docstring, fixed the existing hardcoded test (labels are now compact 1,2,3 instead of raw pixel indices), and added a regression test reproducing this issue's exact repro case (3000×8 float16 image, two disconnected single-pixel regions) confirming they now get distinct nonzero labels.

All existing tests pass across float16/bfloat16/float32/float64 (including gradcheck and torch.jit.script), plus lint/format/doctest are clean. Happy to open the PR once assigned.

**Comment by edgarriba** [comment]
Please send a PR

**Comment by DPug888** [comment]
Hi @ahmadmasood43! Thanks for looking into this. Since I opened the issue and marked it as assigned to myself for a PR, I'm already working on a fix and plan to submit a PR shortly. I appreciate your analysis and effort nonetheless.
